### PR TITLE
Update to WSGI server to fix some issues and add Dualstack

### PR DESCRIPTION
**Pull request by treysis**
* Added dualstack option for webinterface
* Disabled GZip compression by middleware (to prevent Out-Of-Memory errors)

* Updated:
-CherryPy to cheroot/5.11.0
 -> fixes SSL
 -> fixes DualStack
-BottlePy to 0.12.18
-BeautifulSoup to 3.2.0

### CODE REVIEW

File: module/lib/wsgiserver/server.py [hunk 1]
```diff
@@ -0,0 +1,1800 @@
...
+                        # https://github.com/cherrypy/cherrypy/issues/860.
+                        raise
+                else:
+                    # Note that we're explicitly NOT using AI_PASSIVE,
+                    # here, because we want an actual IP to touch.
+                    # localhost won't work if we've bound to a public IP,
+                    # but it will if we bound to '0.0.0.0' (INADDR_ANY).
+                    for res in socket.getaddrinfo(host, port, socket.AF_UNSPEC,
+                                                  socket.SOCK_STREAM):
+                        af, socktype, proto, canonname, sa = res
```
Comment (vuolter) [inline module/lib/wsgiserver/server.py:1733]: Codacy Issue found: [Unused variable 'sa'](https://app.codacy.com/manual/pyLoad/pyload/pullRequest?prid=5459581)

**Comment by vuolter** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 19
- Added 23
           

Complexity increasing per file
==============================
- module/lib/wsgiserver/_compat.py  2
- module/lib/wsgiserver/ssl/pyopenssl.py  15
- module/web/webinterface.py  1
- module/lib/wsgiserver/wsgi.py  8
- module/lib/wsgiserver/workers/threadpool.py  15
- module/lib/wsgiserver/errors.py  3
- module/lib/wsgiserver/ssl/builtin.py  9
- module/lib/wsgiserver/ssl/__init__.py  1
- module/lib/wsgiserver/makefile.py  19
- module/lib/wsgiserver/server.py  35
         

Clones added
============
- module/lib/wsgiserver/makefile.py  2
- module/lib/wsgiserver/server.py  5
         

Clones removed
==============
+ module/lib/wsgiserver/__init__.py  -4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/pyLoad/pyload/pullRequest?prid=5459581&bid=17894560)

File: module/lib/wsgiserver/ssl/pyopenssl.py [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+
+    for f in ('get_context', 'pending', 'send', 'write', 'recv', 'read',
+              'renegotiate', 'bind', 'listen', 'connect', 'accept',
+              'setblocking', 'fileno', 'close', 'get_cipher_list',
+              'getpeername', 'getsockname', 'getsockopt', 'setsockopt',
+              'makefile', 'get_app_data', 'set_app_data', 'state_string',
+              'sock_shutdown', 'get_peer_certificate', 'want_read',
+              'want_write', 'set_connect_state', 'set_accept_state',
+              'connect_ex', 'sendall', 'settimeout', 'gettimeout'):
+        exec("""def %s(self, *args):
```
Comment (vuolter) [inline module/lib/wsgiserver/ssl/pyopenssl.py:134]: Codacy Issue found: [Use of exec](https://app.codacy.com/manual/pyLoad/pyload/pullRequest?prid=5459581)

File: module/lib/wsgiserver/makefile.py [hunk 3]
```diff
@@ -0,0 +1,384 @@
...
+
+    def send(self, data):
+        """Send some part of message to the socket."""
+        bytes_sent = self._sock.send(data)
+        self.bytes_written += bytes_sent
+        return bytes_sent
+
+    def flush(self):
+        """Write all data from buffer to socket and reset write buffer."""
+        if self._wbuf:
```
Comment (vuolter) [inline module/lib/wsgiserver/makefile.py:78]: Codacy Issue found: [Access to member '_wbuf' before its definition line 80](https://app.codacy.com/manual/pyLoad/pyload/pullRequest?prid=5459581)

File: module/lib/wsgiserver/server.py [hunk 4]
```diff
@@ -0,0 +1,1800 @@
...
+
+            # So we can reuse the socket...
+            try:
+                os.unlink(self.bind_addr)
+            except Exception:
+                pass
+
+            # So everyone can access the socket...
+            try:
+                os.chmod(self.bind_addr, 0o777)
```
Comment (vuolter) [inline module/lib/wsgiserver/server.py:1488]: Codacy Issue found: [Chmod setting a permissive mask 0777 on file (NOT PARSED).](https://app.codacy.com/manual/pyLoad/pyload/pullRequest?prid=5459581)

File: module/lib/wsgiserver/server.py [hunk 5]
```diff
@@ -0,0 +1,1800 @@
...
+
+            if k in comma_separated_headers:
+                existing = hdict.get(hname)
+                if existing:
+                    v = b', '.join((existing, v))
+            hdict[hname] = v
+
+        return hdict
+
+    def _allow_header(self, key_name):
```
Comment (vuolter) [inline module/lib/wsgiserver/server.py:167]: Codacy Issue found: [Method could be a function](https://app.codacy.com/manual/pyLoad/pyload/pullRequest?prid=5459581)

File: module/web/utils.py [hunk 6]
```diff
@@ -141,9 +141,12 @@ def toDict(obj):
     return ret
 
 
-class CherryPyWSGI(ServerAdapter):
+class WSGI(ServerAdapter):
     def run(self, handler):
-        from wsgiserver import CherryPyWSGIServer
+        try:
+            from module.lib.wsgiserver.wsgi import Server as WSGIServer
+        except:
```
Comment (vuolter) [inline module/web/utils.py:148]: Codacy Issue found: [No exception type(s) specified](https://app.codacy.com/manual/pyLoad/pyload/pullRequest?prid=5459581)

File: module/lib/wsgiserver/makefile.py [hunk 7]
```diff
@@ -0,0 +1,384 @@
...
+                return ''.join(buffers)
+
+        def readline(self, size=-1):
+            """Read line from the socket to buffer."""
+            data = self._rbuf
+            if size < 0:
+                # Read until \n or EOF, whichever comes first
+                if self._rbufsize <= 1:
+                    # Speed up unbuffered case
+                    assert data == ''
```
Comment (vuolter) [inline module/lib/wsgiserver/makefile.py:318]: Codacy Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/pyLoad/pyload/pullRequest?prid=5459581)

File: module/lib/wsgiserver/makefile.py [hunk 8]
```diff
@@ -0,0 +1,384 @@
...
+                        # - We have no data in our buffer.
+                        # AND
+                        # - Our call to recv returned exactly the
+                        #   number of bytes we were asked to read.
+                        return data
+                    if n == left:
+                        buf.write(data)
+                        del data  # explicit free
+                        break
+                    assert n <= left, 'recv(%d) returned %d bytes' % (left, n)
```
Comment (vuolter) [inline module/lib/wsgiserver/makefile.py:170]: Codacy Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/pyLoad/pyload/pullRequest?prid=5459581)

File: module/lib/wsgiserver/ssl/builtin.py [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+            elif 'handshake operation timed out' in ex.args[0]:
+                # This error is thrown by builtin SSL after a timeout
+                # when client is speaking HTTP to an HTTPS server.
+                # The connection can safely be dropped.
+                return None, {}
+            raise
+        return s, self.get_environ(s)
+
+    # TODO: fill this out more with mod ssl env
+    def get_environ(self, sock):
```
Comment (vuolter) [inline module/lib/wsgiserver/ssl/builtin.py:119]: Codacy Issue found: [Parameters differ from overridden 'get_environ' method](https://app.codacy.com/manual/pyLoad/pyload/pullRequest?prid=5459581)

File: module/lib/wsgiserver/server.py [hunk 10]
```diff
@@ -0,0 +1,1800 @@
...
+
+            invalid_path = (
+                self.strict_mode
+                and not uri.startswith(FORWARD_SLASH)
+                and not uri_is_absolute_form
+            )
+            if invalid_path:
+                # https://tools.ietf.org/html/rfc7230#section-5.3.1
+                # (origin_form) and
+                """Path should start with a forward slash."""
```
Comment (vuolter) [inline module/lib/wsgiserver/server.py:806]: Codacy Issue found: [String statement has no effect](https://app.codacy.com/manual/pyLoad/pyload/pullRequest?prid=5459581)

File: module/lib/wsgiserver/server.py [hunk 11]
```diff
@@ -0,0 +1,1800 @@
...
+            conn = self.ConnectionClass(self, s, mf)
+
+            if not isinstance(self.bind_addr, six.string_types):
+                # optional values
+                # Until we do DNS lookups, omit REMOTE_HOST
+                if addr is None:  # sometimes this can happen
+                    # figure out if AF_INET or AF_INET6.
+                    if len(s.getsockname()) == 2:
+                        # AF_INET
+                        addr = ('0.0.0.0', 0)
```
Comment (vuolter) [inline module/lib/wsgiserver/server.py:1655]: Codacy Issue found: [Possible binding to all interfaces.](https://app.codacy.com/manual/pyLoad/pyload/pullRequest?prid=5459581)

File: module/lib/wsgiserver/server.py [hunk 12]
```diff
@@ -0,0 +1,1800 @@
...
+
+            disallowed_absolute = (
+                self.strict_mode
+                and not self.proxy_mode
+                and uri_is_absolute_form
+            )
+            if disallowed_absolute:
+                # https://tools.ietf.org/html/rfc7230#section-5.3.2
+                # (absolute form)
+                """Absolute URI is only allowed within proxies."""
```
Comment (vuolter) [inline module/lib/wsgiserver/server.py:791]: Codacy Issue found: [String statement has no effect](https://app.codacy.com/manual/pyLoad/pyload/pullRequest?prid=5459581)

File: module/lib/wsgiserver/makefile.py [hunk 13]
```diff
@@ -0,0 +1,384 @@
...
+    def send(self, data):
+        """Send some part of message to the socket."""
+        bytes_sent = self._sock.send(data)
+        self.bytes_written += bytes_sent
+        return bytes_sent
+
+    def flush(self):
+        """Write all data from buffer to socket and reset write buffer."""
+        if self._wbuf:
+            buffer = ''.join(self._wbuf)
```
Comment (vuolter) [inline module/lib/wsgiserver/makefile.py:79]: Codacy Issue found: [Access to member '_wbuf' before its definition line 80](https://app.codacy.com/manual/pyLoad/pyload/pullRequest?prid=5459581)

File: module/lib/wsgiserver/server.py [hunk 14]
```diff
@@ -0,0 +1,1800 @@
...
+            # So we can reuse the socket...
+            try:
+                os.unlink(self.bind_addr)
+            except Exception:
+                pass
+
+            # So everyone can access the socket...
+            try:
+                os.chmod(self.bind_addr, 0o777)
+            except Exception:
```
Comment (vuolter) [inline module/lib/wsgiserver/server.py:1489]: Codacy Issue found: [Try, Except, Pass detected.](https://app.codacy.com/manual/pyLoad/pyload/pullRequest?prid=5459581)

File: module/lib/wsgiserver/server.py [hunk 15]
```diff
@@ -0,0 +1,1800 @@
...
+                if existing:
+                    v = b', '.join((existing, v))
+            hdict[hname] = v
+
+        return hdict
+
+    def _allow_header(self, key_name):
+        return True
+
+    def _transform_key(self, key_name):
```
Comment (vuolter) [inline module/lib/wsgiserver/server.py:170]: Codacy Issue found: [Method could be a function](https://app.codacy.com/manual/pyLoad/pyload/pullRequest?prid=5459581)

File: module/lib/wsgiserver/server.py [hunk 16]
```diff
@@ -0,0 +1,1800 @@
...
+                               traceback=True)
+
+            if self.interrupt:
+                while self.interrupt is True:
+                    # Wait for self.stop() to complete. See _set_interrupt.
+                    time.sleep(0.1)
+                if self.interrupt:
+                    raise self.interrupt
+
+    def error_log(self, msg='', level=20, traceback=False):
```
Comment (vuolter) [inline module/lib/wsgiserver/server.py:1552]: Codacy Issue found: [Method could be a function](https://app.codacy.com/manual/pyLoad/pyload/pullRequest?prid=5459581)

File: module/lib/wsgiserver/server.py [hunk 17]
```diff
@@ -0,0 +1,1800 @@
...
+            # so don't set it.
+            # https://msdn.microsoft.com/en-us/library/ms740621(v=vs.85).aspx
+            self.socket.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
+        if self.nodelay and not isinstance(self.bind_addr, str):
+            self.socket.setsockopt(socket.IPPROTO_TCP, socket.TCP_NODELAY, 1)
+
+        if self.ssl_adapter is not None:
+            self.socket = self.ssl_adapter.bind(self.socket)
+
+        host, port = self.bind_addr[:2]
```
Comment (vuolter) [inline module/lib/wsgiserver/server.py:1583]: Codacy Issue found: [Unused variable 'port'](https://app.codacy.com/manual/pyLoad/pyload/pullRequest?prid=5459581)

File: module/web/ServerThread.py [hunk 18]
```diff
@@ -22,6 +22,8 @@ def __init__(self, pycore):
         self.cert = pycore.config["ssl"]["cert"]
         self.key = pycore.config["ssl"]["key"]
         self.host = pycore.config['webinterface']['host']
+        if self.host == "0.0.0.0" and pycore.config['webinterface']['dualstack']:
```
Comment (vuolter) [inline module/web/ServerThread.py:25]: Codacy Issue found: [Possible binding to all interfaces.](https://app.codacy.com/manual/pyLoad/pyload/pullRequest?prid=5459581)

File: module/lib/wsgiserver/workers/threadpool.py [hunk 19]
```diff
@@ -0,0 +1,267 @@
...
+        self.min = min
+        self.max = max
+        self._threads = []
+        self._queue = queue.Queue(maxsize=accepted_queue_size)
+        self._queue_put_timeout = accepted_queue_timeout
+        self.get = self._queue.get
+
+    def start(self):
+        """Start the pool of threads."""
+        for i in range(self.min):
```
Comment (vuolter) [inline module/lib/wsgiserver/workers/threadpool.py:155]: Codacy Issue found: [Unused variable 'i'](https://app.codacy.com/manual/pyLoad/pyload/pullRequest?prid=5459581)

File: module/lib/wsgiserver/server.py [hunk 20]
```diff
@@ -0,0 +1,1800 @@
...
+                'RFC 7230 (section 3.1.1) and RFC 7231 (section 4.1) '
+                'method names are case-sensitive and uppercase.'
+            )
+            self.simple_response('400 Bad Request', resp)
+            return False
+
+        try:
+            if six.PY2:  # FIXME: Figure out better way to do this
+                # Ref: https://stackoverflow.com/a/196392/595220 (like this?)
+                """This is a dummy check for unicode in URI."""
```
Comment (vuolter) [inline module/lib/wsgiserver/server.py:734]: Codacy Issue found: [String statement has no effect](https://app.codacy.com/manual/pyLoad/pyload/pullRequest?prid=5459581)

File: module/web/webinterface.py [hunk 21]
```diff
@@ -143,17 +143,21 @@ def run_lightweight(host="0.0.0.0", port="8000"):
 
 
 def run_threaded(host="0.0.0.0", port="8000", theads=3, cert="", key=""):
-    from wsgiserver import CherryPyWSGIServer
+    try:
+        from module.lib.wsgiserver.wsgi import Server as WSGIServer
+        from module.lib.wsgiserver.ssl.builtin import BuiltinSSLAdapter
+    except:
```
Comment (vuolter) [inline module/web/webinterface.py:149]: Codacy Issue found: [No exception type(s) specified](https://app.codacy.com/manual/pyLoad/pyload/pullRequest?prid=5459581)

File: module/lib/wsgiserver/ssl/pyopenssl.py (hunk 2 above)
Comment (vuolter) [inline module/lib/wsgiserver/ssl/pyopenssl.py:134]: Codacy Issue found: [Use of exec detected.](https://app.codacy.com/manual/pyLoad/pyload/pullRequest?prid=5459581)

File: module/lib/wsgiserver/workers/threadpool.py [hunk 22]
```diff
@@ -0,0 +1,267 @@
...
+        # calculate the number of threads above the minimum
+        n_extra = max(len(self._threads) - self.min, 0)
+
+        # don't remove more than amount
+        n_to_remove = min(amount, n_extra)
+
+        # put shutdown requests on the queue equal to the number of threads
+        # to remove. As each request is processed by a worker, that worker
+        # will terminate and be culled from the list.
+        for n in range(n_to_remove):
```
Comment (vuolter) [inline module/lib/wsgiserver/workers/threadpool.py:219]: Codacy Issue found: [Unused variable 'n'](https://app.codacy.com/manual/pyLoad/pyload/pullRequest?prid=5459581)

File: module/web/utils.py (hunk 6 above)
Comment (GammaC0de) [inline module/web/utils.py:148]: This should be: `except ImportError`, not generic execpt

File: module/web/webinterface.py [hunk 23]
```diff
@@ -124,7 +124,7 @@
 }
 
 web = StripPathMiddleware(SessionMiddleware(app(), session_opts))
-web = GZipMiddleWare(web)
+#web = GZipMiddleWare(web)
```
Comment (GammaC0de) [inline module/web/webinterface.py:127]: Please uncomment, we will deal with the downloading issue (#3541) later.

File: module/web/webinterface.py (hunk 21 above)
Comment (GammaC0de) [inline module/web/webinterface.py:149]: Again, `execpt ImportError:`

**Comment by treysis** [comment]
So do I have to make a new commit to my fork and then issue a new pull-request? I am quite new to git...

**Comment by GammaC0de** [comment]
Before we merge this, 
Can you please explain what is `dualstack ` in words that a common folk can understand.

**Comment by treysis** [comment]
Dualstack means webinterface is both listening on IPv4 and IPv6. So, if you have an IPv6-only connection, you can access it from both IPv4 and IPv6 connections. CherryPy/cheroot will listen by default on both protocols if listening address is set to `"::"`. `"0.0.0.0"` will force it to listen only on IPv4. Unfortunately, it is not possible to use `"::"` as IP address in `pyload.conf` due to the way the parsing algorithm works. Therefor I implemented this workaround:

If you set the listening address to `0.0.0.0` and turn dualstack option on, it will set the listening address to `::` and listen on both IPv4 and IPv6. If you set it to off it will listen only on `0.0.0.0` or on any address (incl. IPv6) that you specify.

**Comment by vuolter** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 19
- Added 21
           

Complexity increasing per file
==============================
- module/lib/wsgiserver/makefile.py  19
- module/lib/wsgiserver/ssl/__init__.py  1
- module/lib/wsgiserver/server.py  35
- module/lib/wsgiserver/_compat.py  2
- module/lib/wsgiserver/workers/threadpool.py  15
- module/lib/wsgiserver/wsgi.py  8
- module/lib/wsgiserver/ssl/pyopenssl.py  15
- module/lib/wsgiserver/errors.py  3
- module/web/webinterface.py  1
- module/lib/wsgiserver/ssl/builtin.py  9
         

Clones added
============
- module/lib/wsgiserver/makefile.py  2
- module/lib/wsgiserver/server.py  5
         

Clones removed
==============
+ module/lib/wsgiserver/__init__.py  -4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/pyLoad/pyload/pullRequest?prid=5459581&bid=17894560)

File: module/lib/wsgiserver/server.py (hunk 15 above)
Comment (vuolter) [inline module/lib/wsgiserver/server.py:170]: Codacy Issue found: [Method could be a function](https://app.codacy.com/manual/pyLoad/pyload/pullRequest?prid=5459581)

File: module/lib/wsgiserver/ssl/pyopenssl.py (hunk 2 above)
Comment (vuolter) [inline module/lib/wsgiserver/ssl/pyopenssl.py:134]: Codacy Issue found: [Use of exec](https://app.codacy.com/manual/pyLoad/pyload/pullRequest?prid=5459581)

Comment (vuolter) [inline module/lib/wsgiserver/ssl/pyopenssl.py:134]: Codacy Issue found: [Use of exec detected.](https://app.codacy.com/manual/pyLoad/pyload/pullRequest?prid=5459581)

File: module/lib/wsgiserver/makefile.py (hunk 8 above)
Comment (vuolter) [inline module/lib/wsgiserver/makefile.py:170]: Codacy Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/pyLoad/pyload/pullRequest?prid=5459581)

File: module/lib/wsgiserver/ssl/builtin.py (hunk 9 above)
Comment (vuolter) [inline module/lib/wsgiserver/ssl/builtin.py:119]: Codacy Issue found: [Parameters differ from overridden 'get_environ' method](https://app.codacy.com/manual/pyLoad/pyload/pullRequest?prid=5459581)

File: module/lib/wsgiserver/server.py (hunk 5 above)
Comment (vuolter) [inline module/lib/wsgiserver/server.py:167]: Codacy Issue found: [Method could be a function](https://app.codacy.com/manual/pyLoad/pyload/pullRequest?prid=5459581)

File: module/lib/wsgiserver/server.py (hunk 16 above)
Comment (vuolter) [inline module/lib/wsgiserver/server.py:1552]: Codacy Issue found: [Method could be a function](https://app.codacy.com/manual/pyLoad/pyload/pullRequest?prid=5459581)

File: module/lib/wsgiserver/server.py (hunk 11 above)
Comment (vuolter) [inline module/lib/wsgiserver/server.py:1655]: Codacy Issue found: [Possible binding to all interfaces.](https://app.codacy.com/manual/pyLoad/pyload/pullRequest?prid=5459581)

File: module/lib/wsgiserver/server.py (hunk 1 above)
Comment (vuolter) [inline module/lib/wsgiserver/server.py:1733]: Codacy Issue found: [Unused variable 'sa'](https://app.codacy.com/manual/pyLoad/pyload/pullRequest?prid=5459581)

File: module/lib/wsgiserver/makefile.py (hunk 13 above)
Comment (vuolter) [inline module/lib/wsgiserver/makefile.py:79]: Codacy Issue found: [Access to member '_wbuf' before its definition line 80](https://app.codacy.com/manual/pyLoad/pyload/pullRequest?prid=5459581)

File: module/lib/wsgiserver/server.py (hunk 10 above)
Comment (vuolter) [inline module/lib/wsgiserver/server.py:806]: Codacy Issue found: [String statement has no effect](https://app.codacy.com/manual/pyLoad/pyload/pullRequest?prid=5459581)

File: module/lib/wsgiserver/makefile.py (hunk 7 above)
Comment (vuolter) [inline module/lib/wsgiserver/makefile.py:318]: Codacy Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/pyLoad/pyload/pullRequest?prid=5459581)

File: module/lib/wsgiserver/server.py (hunk 12 above)
Comment (vuolter) [inline module/lib/wsgiserver/server.py:791]: Codacy Issue found: [String statement has no effect](https://app.codacy.com/manual/pyLoad/pyload/pullRequest?prid=5459581)

File: module/lib/wsgiserver/server.py (hunk 17 above)
Comment (vuolter) [inline module/lib/wsgiserver/server.py:1583]: Codacy Issue found: [Unused variable 'port'](https://app.codacy.com/manual/pyLoad/pyload/pullRequest?prid=5459581)

File: module/web/ServerThread.py (hunk 18 above)
Comment (vuolter) [inline module/web/ServerThread.py:25]: Codacy Issue found: [Possible binding to all interfaces.](https://app.codacy.com/manual/pyLoad/pyload/pullRequest?prid=5459581)

File: module/lib/wsgiserver/makefile.py (hunk 3 above)
Comment (vuolter) [inline module/lib/wsgiserver/makefile.py:78]: Codacy Issue found: [Access to member '_wbuf' before its definition line 80](https://app.codacy.com/manual/pyLoad/pyload/pullRequest?prid=5459581)

File: module/lib/wsgiserver/server.py (hunk 4 above)
Comment (vuolter) [inline module/lib/wsgiserver/server.py:1488]: Codacy Issue found: [Chmod setting a permissive mask 0777 on file (NOT PARSED).](https://app.codacy.com/manual/pyLoad/pyload/pullRequest?prid=5459581)

File: module/lib/wsgiserver/server.py (hunk 14 above)
Comment (vuolter) [inline module/lib/wsgiserver/server.py:1489]: Codacy Issue found: [Try, Except, Pass detected.](https://app.codacy.com/manual/pyLoad/pyload/pullRequest?prid=5459581)

File: module/lib/wsgiserver/workers/threadpool.py (hunk 22 above)
Comment (vuolter) [inline module/lib/wsgiserver/workers/threadpool.py:219]: Codacy Issue found: [Unused variable 'n'](https://app.codacy.com/manual/pyLoad/pyload/pullRequest?prid=5459581)

File: module/lib/wsgiserver/server.py (hunk 20 above)
Comment (vuolter) [inline module/lib/wsgiserver/server.py:734]: Codacy Issue found: [String statement has no effect](https://app.codacy.com/manual/pyLoad/pyload/pullRequest?prid=5459581)

File: module/lib/wsgiserver/workers/threadpool.py (hunk 19 above)
Comment (vuolter) [inline module/lib/wsgiserver/workers/threadpool.py:155]: Codacy Issue found: [Unused variable 'i'](https://app.codacy.com/manual/pyLoad/pyload/pullRequest?prid=5459581)

File: module/config/default.conf [hunk 24]
```diff
@@ -13,6 +13,7 @@ webinterface - "Webinterface":
 	bool activated : "Activated" = True
 	builtin;threaded;fastcgi;lightweight server : "Server" = builtin
 	bool https : "Use HTTPS" = False
+	bool dualstack : "Use Dualstack" = True
```
Comment (GammaC0de) [inline module/config/default.conf:16]: Please change to `bool dualstack : "Listen on both IPv4 and IPv6 (IP must be set to 0.0.0.0)" = True`

**Comment by vuolter** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 19
- Added 21
           

Complexity increasing per file
==============================
- module/web/webinterface.py  1
- module/lib/wsgiserver/server.py  35
- module/lib/wsgiserver/makefile.py  19
- module/lib/wsgiserver/ssl/__init__.py  1
- module/lib/wsgiserver/_compat.py  2
- module/lib/wsgiserver/workers/threadpool.py  15
- module/lib/wsgiserver/errors.py  3
- module/lib/wsgiserver/ssl/builtin.py  9
- module/lib/wsgiserver/wsgi.py  8
- module/lib/wsgiserver/ssl/pyopenssl.py  15
         

Clones added
============
- module/lib/wsgiserver/server.py  5
- module/lib/wsgiserver/makefile.py  2
         

Clones removed
==============
+ module/lib/wsgiserver/__init__.py  -4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/pyLoad/pyload/pullRequest?prid=5459581&bid=17894560)

File: module/lib/wsgiserver/ssl/pyopenssl.py (hunk 2 above)
Comment (vuolter) [inline module/lib/wsgiserver/ssl/pyopenssl.py:134]: Codacy Issue found: [Use of exec detected.](https://app.codacy.com/manual/pyLoad/pyload/pullRequest?prid=5459581)

File: module/lib/wsgiserver/server.py (hunk 12 above)
Comment (vuolter) [inline module/lib/wsgiserver/server.py:791]: Codacy Issue found: [String statement has no effect](https://app.codacy.com/manual/pyLoad/pyload/pullRequest?prid=5459581)

File: module/lib/wsgiserver/makefile.py (hunk 7 above)
Comment (vuolter) [inline module/lib/wsgiserver/makefile.py:318]: Codacy Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/pyLoad/pyload/pullRequest?prid=5459581)

File: module/web/ServerThread.py (hunk 18 above)
Comment (vuolter) [inline module/web/ServerThread.py:25]: Codacy Issue found: [Possible binding to all interfaces.](https://app.codacy.com/manual/pyLoad/pyload/pullRequest?prid=5459581)

File: module/lib/wsgiserver/server.py (hunk 15 above)
Comment (vuolter) [inline module/lib/wsgiserver/server.py:170]: Codacy Issue found: [Method could be a function](https://app.codacy.com/manual/pyLoad/pyload/pullRequest?prid=5459581)

File: module/lib/wsgiserver/server.py (hunk 17 above)
Comment (vuolter) [inline module/lib/wsgiserver/server.py:1583]: Codacy Issue found: [Unused variable 'port'](https://app.codacy.com/manual/pyLoad/pyload/pullRequest?prid=5459581)

File: module/lib/wsgiserver/makefile.py (hunk 13 above)
Comment (vuolter) [inline module/lib/wsgiserver/makefile.py:79]: Codacy Issue found: [Access to member '_wbuf' before its definition line 80](https://app.codacy.com/manual/pyLoad/pyload/pullRequest?prid=5459581)

File: module/lib/wsgiserver/server.py (hunk 14 above)
Comment (vuolter) [inline module/lib/wsgiserver/server.py:1489]: Codacy Issue found: [Try, Except, Pass detected.](https://app.codacy.com/manual/pyLoad/pyload/pullRequest?prid=5459581)

File: module/lib/wsgiserver/ssl/builtin.py (hunk 9 above)
Comment (vuolter) [inline module/lib/wsgiserver/ssl/builtin.py:119]: Codacy Issue found: [Parameters differ from overridden 'get_environ' method](https://app.codacy.com/manual/pyLoad/pyload/pullRequest?prid=5459581)

File: module/lib/wsgiserver/workers/threadpool.py (hunk 22 above)
Comment (vuolter) [inline module/lib/wsgiserver/workers/threadpool.py:219]: Codacy Issue found: [Unused variable 'n'](https://app.codacy.com/manual/pyLoad/pyload/pullRequest?prid=5459581)

File: module/lib/wsgiserver/makefile.py (hunk 3 above)
Comment (vuolter) [inline module/lib/wsgiserver/makefile.py:78]: Codacy Issue found: [Access to member '_wbuf' before its definition line 80](https://app.codacy.com/manual/pyLoad/pyload/pullRequest?prid=5459581)

File: module/lib/wsgiserver/server.py (hunk 1 above)
Comment (vuolter) [inline module/lib/wsgiserver/server.py:1733]: Codacy Issue found: [Unused variable 'sa'](https://app.codacy.com/manual/pyLoad/pyload/pullRequest?prid=5459581)

File: module/lib/wsgiserver/makefile.py (hunk 8 above)
Comment (vuolter) [inline module/lib/wsgiserver/makefile.py:170]: Codacy Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/pyLoad/pyload/pullRequest?prid=5459581)

File: module/lib/wsgiserver/server.py (hunk 5 above)
Comment (vuolter) [inline module/lib/wsgiserver/server.py:167]: Codacy Issue found: [Method could be a function](https://app.codacy.com/manual/pyLoad/pyload/pullRequest?prid=5459581)

File: module/lib/wsgiserver/server.py (hunk 16 above)
Comment (vuolter) [inline module/lib/wsgiserver/server.py:1552]: Codacy Issue found: [Method could be a function](https://app.codacy.com/manual/pyLoad/pyload/pullRequest?prid=5459581)

File: module/lib/wsgiserver/server.py (hunk 11 above)
Comment (vuolter) [inline module/lib/wsgiserver/server.py:1655]: Codacy Issue found: [Possible binding to all interfaces.](https://app.codacy.com/manual/pyLoad/pyload/pullRequest?prid=5459581)

File: module/lib/wsgiserver/workers/threadpool.py (hunk 19 above)
Comment (vuolter) [inline module/lib/wsgiserver/workers/threadpool.py:155]: Codacy Issue found: [Unused variable 'i'](https://app.codacy.com/manual/pyLoad/pyload/pullRequest?prid=5459581)

File: module/lib/wsgiserver/ssl/pyopenssl.py (hunk 2 above)
Comment (vuolter) [inline module/lib/wsgiserver/ssl/pyopenssl.py:134]: Codacy Issue found: [Use of exec](https://app.codacy.com/manual/pyLoad/pyload/pullRequest?prid=5459581)

File: module/lib/wsgiserver/server.py (hunk 10 above)
Comment (vuolter) [inline module/lib/wsgiserver/server.py:806]: Codacy Issue found: [String statement has no effect](https://app.codacy.com/manual/pyLoad/pyload/pullRequest?prid=5459581)

File: module/lib/wsgiserver/server.py (hunk 20 above)
Comment (vuolter) [inline module/lib/wsgiserver/server.py:734]: Codacy Issue found: [String statement has no effect](https://app.codacy.com/manual/pyLoad/pyload/pullRequest?prid=5459581)

File: module/lib/wsgiserver/server.py (hunk 4 above)
Comment (vuolter) [inline module/lib/wsgiserver/server.py:1488]: Codacy Issue found: [Chmod setting a permissive mask 0777 on file (NOT PARSED).](https://app.codacy.com/manual/pyLoad/pyload/pullRequest?prid=5459581)

**Comment by GammaC0de** [comment]
👍 Thanks for your contribution.

**Pull request merged**